### ISSUE TITLE
AttachToContainer hangs or returns prematurely

### ISSUE THREAD

**Issue by pwaller** [fsouza/go-dockerclient#415]
This is a followup to what I posted in #209.

I've [written a minimal reproducer](https://gist.github.com/pwaller/314c7591a8bfd32f8999/fabc0320618d9170426e890611c7131c10baa980) which reliably hits the problem.

Grab it:
```
$ go get -v gist.github.com/314c7591a8bfd32f8999.git
```

There are two problems. The first is that the output is reliably not the correct length:

```
$ 314c7591a8bfd32f8999.git debian:wheezy cat /bin/bash | wc -c
41051
$ docker run --rm debian:wheezy cat /bin/bash | wc -c
975488
```

Secondly, sometimes it just freezes mid stream:

```
$ while sleep 0.1; do 314c7591a8bfd32f8999.git debian:wheezy cat /bin/bash | wc -c; done
[runs a few times before hanging]
```

Note: I've reproduced this on docker 1.7.1 (reliably returns zero bytes) 1.8.3 (hangs sometimes, reliably returns 41,051 bytes) and 1.9.0 (hangs more frequently, I've seen it return 41,051 bytes and also 1,519 bytes).

**Comment by pwaller** [comment]
Okay, I think I've cracked the problem. This diff appears to make it work reliably:

```diff
diff --git a/dockerclient-bug.go b/dockerclient-bug.go
index f394681..b045e71 100644
--- a/main.go
+++ b/main.go
@@ -51,12 +51,7 @@ func DockerRun(c *docker.Client, imageName string, args []string) (io.ReadCloser
                return nil, err
        }
 
-       log.Printf("Start container...")
-       err = c.StartContainer(cont.ID, &docker.HostConfig{})
-       if err != nil {
-               log.Printf("Start container... failed: %v", err)
-               return nil, err
-       }
+       attached := make(chan struct{})
 
        r, w := io.Pipe()
 
@@ -70,12 +65,23 @@ func DockerRun(c *docker.Client, imageName string, args []string) (io.ReadCloser
                        Stdout:       true,
                        Stderr:       true,
                        Stream:       true,
+                       Success:      attached,
                })
                log.Printf("~AttachToContainer")
                // io.Pipe hardwired to never return error here.
                _ = w.CloseWithError(err)
        }()
 
+       <-attached
+       attached <- struct{}{}
+
+       log.Printf("Start container...")
+       err = c.StartContainer(cont.ID, &docker.HostConfig{})
+       if err != nil {
+               log.Printf("Start container... failed: %v", err)
+               return nil, err
+       }
+
        removeContainer := func() {
                err := c.RemoveContainer(docker.RemoveContainerOptions{
                        ID:            cont.ID,
```

Then there is a second problem: if StartContainer fails, there is no way I can find to cause the `AttachToContainer` call to return, leading to leaked resources. I guess that's already covered well in this bug? https://github.com/fsouza/go-dockerclient/issues/161

**Comment by fsouza** [comment]
Yeah, you would be able to detect that start failed and then cancel AttachToContainer.

Sounds like we definitely need to tackle #161 :-(

**Comment by pwaller** [comment]
OK. I'm going to close this one in favour of #161.

**Comment by pwaller** [comment]
Renamed since probably it was not the "binary stream" that caused it, but just that my stream was large enough/being transmitted fast enough compared to all of the text streams I tested.

**Comment by pwaller** [comment]
@fsouza is there something to be addressed here which is not #161?

**Comment by fsouza** [comment]
@pwaller I don't think so, sorry for reopening the issue. Adding some more information, as far as I remember, this is the way Docker guarantees getting all the output from a container: attach to the container before starting it.

Thanks for bringing some light to the issue.

**Comment by pwaller** [comment]
:+1:. Any chance we can start adding a context to `AttachToContainerOptions`, etc?

**Comment by fsouza** [comment]
@pwaller that's what I plan to do, but unfortunately I don't have an ETA. Would you be interested in working on that?

**Comment by pwaller** [comment]
If I get a few spare cycles I'll take a shot at it - no promises or ETA from me either. Feel free to ping me if/when it comes time for review though.

**Comment by nkovacs** [comment]
> Then there is a second problem: if StartContainer fails, there is no way I can find to cause the AttachToContainer call to return, leading to leaked resources. I guess that's already covered well in this bug?

Are you sure? `AttachToContainer` succeeds before `StartContainer` is called, so the goroutine shouldn't leak.
